### PR TITLE
feat(ibc): impl ibc gRPC query and msg trait

### PR DESCRIPTION
**Pull request by wenyuanhust**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Which docs this PR relation**:

Ref #

**Which toolchain this PR adaption**:

No Breaking Change

**Special notes for your reviewer**:

### CI Switch

ci-runs-only: [Chaos CI, Cargo Clippy, Coverage Test, E2E Tests, Code Format, Unit Tests, Web3 Compatible Tests]

### CODE REVIEW

File: core/ibc/src/grpc.rs [hunk 1]
```diff
@@ -1 +1,913 @@
...
+            .add_service(ibc_conn_service)
+            .add_service(ibc_channel_service)
+            .add_service(ibc_client_msg_service)
+            .serve(self.addr)
+            .await
+            .unwrap();
+    }
+
+    pub fn client_service(&self) -> ClientQueryServer<IbcClientService<Adapter>> {
+        ClientQueryServer::new(IbcClientService::new(self.store.clone()))
```
Comment (KaoImin) [inline core/ibc/src/grpc.rs]:
```suggestion
        ClientQueryServer::new(IbcClientService::new(Arc::clone(&self.store))
```

File: core/ibc/src/lib.rs [hunk 2]
```diff
@@ -35,10 +35,23 @@ use ibc::{
     Height,
 };
 
+use protocol::traits::IbcAdapter;
```
Comment (KaoImin) [inline core/ibc/src/lib.rs]:
```suggestion
use protocol::traits::{IbcAdapter, IbcContext};
```

File: core/ibc/src/lib.rs [hunk 3]
```diff
@@ -35,10 +35,23 @@ use ibc::{
...
 use protocol::types::Hasher;
 
 use protocol::traits::IbcContext;
 
+use crate::grpc::GrpcService;
+
+pub struct GrpcAdapter {}
+impl IbcAdapter for GrpcAdapter {}
+
+pub async fn run_ibc_grpc(addr: String) {
```
Comment (KaoImin) [inline core/ibc/src/lib.rs]:
```suggestion
pub async fn run_ibc_grpc<Adapter: IbcAdapter>(adapter: Adapter, addr: String) {
```

File: protocol/src/traits/ibc.rs [hunk 4]
```diff
@@ -1,14 +1,183 @@
...
 use crate::traits::IbcCrossChainStorage;
 use crate::types::{ConsensusStateWithHeight, Header, Metadata};
 use crate::{async_trait, ProtocolResult};
+use std::{
+    fmt::{Display, Formatter},
+    ops::Deref,
+    str::{from_utf8, FromStr},
+};
+#[derive(Debug, Ord, PartialOrd, Eq, PartialEq, Clone)]
+pub struct Error(String);
```
Comment (KaoImin) [inline protocol/src/traits/ibc.rs]: Where is this error used?

File: protocol/src/traits/ibc.rs [hunk 5]
```diff
@@ -1,14 +1,183 @@
...
+/// Block height
+pub(crate) type RawHeight = u64;
+
+/// Store height to query
+#[derive(Debug, Copy, Clone)]
+pub enum Height {
+    Pending,
+    // Latest,
+    // Stable(RawHeight), // or equivalently `tendermint::block::Height`
+}
```
Comment (KaoImin) [inline protocol/src/traits/ibc.rs]: The `Identity` and `Path` should be defined in `protocol/src/types/ibc.rs`.

File: protocol/src/traits/ibc.rs (hunk 4 above)
Comment (wenyuanhust) [reply protocol/src/traits/ibc.rs]: just to simplify the deleted `flex-error`, used by `Path`

File: core/ibc/src/lib.rs [hunk 6]
```diff
@@ -35,9 +35,23 @@ use ibc::{
...
 
+use protocol::traits::{IbcAdapter, IbcContext};
 use protocol::types::Hasher;
 
-use protocol::traits::IbcContext;
+use crate::grpc::GrpcService;
+
+pub async fn run_ibc_grpc<
+    Adapter: IbcAdapter + 'static,
+    Ctx: Ics26Context + Sync + Send + 'static,
```
Comment (KaoImin) [inline core/ibc/src/lib.rs:45]: The `Ctx` is bound in `Arc<Rwlock<Ctx>>` so the `Sync + Send` is useless.

**Review by KaoImin — APPROVED** [review]

**Pull request merged**